src/basic/os-util.c

```diff
@@ -67,6 +70,63 @@ int open_extension_release(const char *root, const char *extension, char **ret_p
                 r = chase_symlinks(extension_full_path, root, CHASE_PREFIX_ROOT,
                                    ret_path ? &q : NULL,
                                    ret_fd ? &fd : NULL);
+                /* Cannot find the expected extension-release file? The image filename might have been
+                 * mangled on deployment, so fallback to checking for any file in the extension-release.d
+                 * directory, and return the first one with a user.extension-release xattr instead.
+                 * The user.extension-release.strict xattr is checked to ensure the author of the image
+                 * considers it OK if names do not match. */
+                if (r == -ENOENT) {
+                        _cleanup_free_ char *extension_release_dir_path = NULL;
+                        _cleanup_closedir_ DIR *extension_release_dir = NULL;
+
+                        r = chase_symlinks_and_opendir("/usr/lib/extension-release.d/", root, CHASE_PREFIX_ROOT,
+                                                       &extension_release_dir_path, &extension_release_dir);
+                        if (r < 0)
+                                return r;
+
+                        r = -ENOENT;
+                        struct dirent *de;
+                        FOREACH_DIRENT(de, extension_release_dir, return -errno) {
+                                if (!IN_SET(de->d_type, DT_REG, DT_UNKNOWN))
+                                        continue;
+
+                                const char *image_name = startswith(de->d_name, "extension-release.");
+                                if (!image_name)
+                                        continue;
+
+                                if (!image_name_is_valid(image_name))
+                                        continue;
+
+                                /* We already chased the directory, and checked that
+                                 * this is a real file, so we shouldn't fail to open it. */
+                                _cleanup_close_ int extension_release_fd = openat(dirfd(extension_release_dir), de->d_name, O_RDONLY|O_CLOEXEC|O_NOCTTY|O_NOFOLLOW);
```

Comment:
I'm getting EBADF here if I use `O_PATH`

Reply:
we use it like that all over the place.

note that the flags combo must be more reduced than without O_PATH, i.e. you need to drop O_NOCTTY and O_RDONLY.

Reply:
i.e. you want O_PATH|O_CLOEXEC|O_NOFOLLOW

Reply:
but opening this as a file is probably OK too.

Reply:
Ah, the problem is that you can't read an extended attribute out of a O_PATH file descriptor, you get EBADF. I'll leave as-is.

Reply:
oh, you can, see how fgetxattr_fake() does it. It uses path based getxattr() but then uses /proc/self/fd/… for the O_PATH fd.

Reply:
super long line, maybe break it somehwere?

Reply:
alright, added a fgetxattr_fake_malloc helper for this